src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPECompilerManager.cs

```diff
@@ -0,0 +1,372 @@
+        {
+            return Path.Combine(projectPath, intermediateOutputPath, "TempPE");
+        }
+
+        private void QueueCompilation()
+        {
+            if (_filesToCompile.Count > 0)
+            {
+                _scheduler.ScheduleAsyncTask(ProcessCompileQueue, _project.Services.ProjectAsynchronousTasks.UnloadCancellationToken);
+            }
+        }
+
+        protected override async Task DisposeCoreAsync(bool initialized)
+        {
+            if (initialized)
+            {
+                // This will stop our blocks taking any more input
+                _inputsActionBlock?.Complete();
+                _fileWatcherActionBlock?.Complete();
+
+                if (_inputsActionBlock != null)
+                {
+                    await Task.WhenAll(_inputsActionBlock.Completion, _fileWatcherActionBlock!.Completion);
+                }
+
+                // By waiting for completion we know that the following dispose will cancel any pending compilations, and there won't be any more
+                _scheduler.Dispose();
+                _disposables.Dispose();
+            }
+        }
+
+        private async Task ProcessCompileQueue(CancellationToken token)
+        {
+            await _activeWorkspaceProjectContextHost.OpenContextForWriteAsync(async accessor =>
+            {
+                // Grab the next file to compile off the queue
+                (string fileName, bool ignoreFileWriteTime) = _filesToCompile.FirstOrDefault();
+                while (fileName != null)
+                {
+                    if (IsDisposing || IsDisposed)
+                    {
+                        return;
+                    }
+
+                    token.ThrowIfCancellationRequested();
+
+                    // Remove the file from our todo list. If it wasn't there (because it was removed as a design time input while we were busy) we don't need to compile it
+                    bool wasInQueue = ThreadingTools.ApplyChangeOptimistically(ref _filesToCompile, fileName, (s, f) => s.Remove(f));
+
+                    string outputFileName = await GetOutputFileName(fileName);
+                    if (wasInQueue)
+                    {
+                        await CompileDesignTimeInput(accessor.Context, fileName, outputFileName, ignoreFileWriteTime, token);
+                    }
+
+                    // Grab another file off the queue
+                    (fileName, ignoreFileWriteTime) = _filesToCompile.FirstOrDefault();
+                }
+            });
+        }
+
+        /// <summary>
+        /// Gets the XML that describes a TempPE DLL, including building it if necessary
+        /// </summary>
+        /// <param name="fileName"></param>
+        /// <returns></returns>
+        public async Task<string> GetDesignTimeInputXML(string fileName)
+        {
+            // Make sure we're not being asked to compile a random file
```

Comment:
I think you wanted to initialize here?

How does the caller know when they can call this method?

Reply:
As above, this is still to come.